File: macos/Services/MCP/Tools/Tier3/DeleteRowsTool.swift
// DeleteRowsTool.swift
// Gridex
//
// MCP Tool: Delete rows from a table. Requires user approval and WHERE clause.

import Foundation

struct DeleteRowsTool: MCPTool {
    let name = "delete_rows"
    let description = "Delete rows matching WHERE clause. Requires user approval. WHERE clause is MANDATORY."
    let tier = MCPPermissionTier.write

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "table_name": [
                "type": "string",
                "description": "Name of the table to delete from (letters, digits, underscore only)"
            ],
            "schema": [
                "type": "string",
                "description": "Optional schema name (letters, digits, underscore only)"
            ],
            "where": [
                "type": "string",
                "description": "WHERE clause (required). Must not contain ';', '--', '/*', or '*/'."
            ],
            "where_params": [
                "type": "array",
                "description": "Parameters for WHERE clause placeholders"
            ]
        ],
        "required": ["connection_id", "table_name", "where"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)
        let (tableName, schemaName) = try MCPIdentifierValidator.extractTableAndSchema(from: params)

        guard let whereClause = params["where"]?.stringValue else {
            throw MCPToolError.invalidParameters("where clause is required. Bare DELETE without WHERE is not allowed.")
        }

        if let error = await context.permissionEngine.validateWhereClause(whereClause).errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let (adapter, config) = try await context.getAdapter(for: connectionId)
        let qualifiedTable = config.databaseType.sqlDialect.qualifiedIdentifier(tableName, schema: schemaName)
        let deleteSQL = "DELETE FROM \(qualifiedTable) WHERE \(whereClause)"

        let estimatedRows = await MCPRowCountEstimator.estimate(
            adapter: adapter,
            qualifiedTable: qualifiedTable,
            whereClause: whereClause,
            config: config
        )

        if permission.requiresUserApproval {
            let details = """
            SQL: \(deleteSQL)

            Estimated rows to delete: \(estimatedRows)

            ⚠️ This operation cannot be undone!
            """

            let approved = await context.requestApproval(
                tool: name,
                description: "Delete rows from '\(tableName)' where \(whereClause)",
                details: details,
                connectionId: connectionId
            )
            if !approved {
                throw MCPToolError.permissionDenied("User denied the operation.")
            }
        }

        let result = try await adapter.executeRaw(sql: deleteSQL)
        return MCPToolResult(text: "Successfully deleted \(result.rowsAffected) row(s) from '\(tableName)'.")
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier3/DeleteRowsTool.cpp
//
// DeleteRowsTool.cpp
//

#include "DeleteRowsTool.h"
#include "../MCPToolHelpers.h"
#include "../../Security/MCPIdentifierValidator.h"
#include "../../Security/MCPRowCountEstimator.h"

namespace DBModels
{
    MCPToolResult DeleteRowsTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);
        auto ts = MCPIdentifierValidator::extractTableAndSchema(params);

        if (!params.contains("where") || !params["where"].is_string())
            throw MCPToolError::invalidParameters(
                "where is required. Bare DELETE without WHERE is not allowed.");
        const auto whereUtf8 = params["where"].get<std::string>();

        {
            const auto v = ctx.permissionEngine.validateWhereClause(whereUtf8);
            if (const auto* msg = v.errorMessage())
                throw MCPToolError::permissionDenied(*msg);
        }

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, config] = ctx.getAdapter(connId);
        const std::wstring tableW = MCPToolHelpers::fromUtf8(ts.table);
        const std::wstring schemaW = ts.schema.has_value()
            ? MCPToolHelpers::fromUtf8(*ts.schema) : std::wstring{};
        const std::wstring whereW = MCPToolHelpers::fromUtf8(whereUtf8);

        std::wstring qTable;
        try
        {
            qTable = (schemaW.empty())
                ? adapter->quoteSqlIdentifier(tableW)
                : adapter->quoteSqlIdentifier(schemaW) + L"." + adapter->quoteSqlIdentifier(tableW);
        }
        catch (...) { qTable = tableW; }

        const std::wstring sql = L"DELETE FROM " + qTable + L" WHERE " + whereW;

        if (perm.requiresUserApproval())
        {
            const int est = MCPRowCountEstimator::estimate(
                adapter, qTable, whereW, config);
            const std::wstring details =
                L"SQL:\n" + sql +
                L"\n\nEstimated rows to delete: " + std::to_wstring(est) +
                L"\n\n⚠ This operation cannot be undone!";
            const auto clientW = MCPToolHelpers::fromUtf8(
                ctx.client.name.empty() ? std::string{"AI client"} : ctx.client.name);
            const bool ok = ctx.requestApproval(
                MCPToolHelpers::fromUtf8(this->name()),
                L"Delete rows from '" + tableW + L"' where " + whereW,
                details, connId, clientW);
            if (!ok) throw MCPToolError::permissionDenied("User denied the operation.");
        }

        QueryResult r;
        try { r = adapter->execute(sql); }
        catch (const std::exception& e)
        { throw MCPToolError::queryFailed(e.what()); }
        if (!r.success) throw MCPToolError::queryFailed(MCPToolHelpers::toUtf8(r.error));

        return MCPToolResult::text(
            "Deleted " + std::to_string(r.totalRows) + " row(s) from '" + ts.table + "'.");
    }
}
